atlasdb-commons/src/main/java/com/palantir/async/initializer/AsyncInitializer.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc. All rights reserved.
+ *
+ * Licensed under the BSD-3 License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.async.initializer;
+
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.palantir.common.concurrent.NamedThreadFactory;
+import com.palantir.exception.NotInitializedException;
+import com.palantir.logsafe.SafeArg;
+
+/**
+ * Implements basic infrastructure to allow an object to be asynchronously initialized.
+ * In order to be ThreadSafe, the abstract methods of the inheriting class need to be synchronized.
+ */
+@ThreadSafe
+public abstract class AsyncInitializer {
+    private static final Logger log = LoggerFactory.getLogger(AsyncInitializer.class);
+
+    private final ScheduledExecutorService singleThreadedExecutor = getExecutorService();
+
+    private final AtomicBoolean isInitializing = new AtomicBoolean(false);
+    private volatile boolean initialized = false;
+    private volatile boolean canceledInitialization = false;
+
+    public final void initialize(boolean initializeAsync) {
+        assertNeverCalledInitialize();
+
+        if (!initializeAsync) {
+            tryInitializeInternal();
+            return;
+        }
+
+        try {
+            tryInitializeInternal();
```

Comment:
Since this isn't on the executor, it sort of breaks the guarantees of the `cancelInitialization` method (i.e. you could call `cancelInitialization` while this is running and the callback would never get executed. Are we ok with that?

Also this try/catch/cleanup logic is duplicated in `scheduleInitialization` - may want to refactor a method

Reply:
The initialize method is called on the static `create` methods of each class been async initialized. This means that 1. `close` cannot be called, since the object is not out of the scope of the static method, and 2. we preserve the original 'synchronous' behavior if possible, even if the the async flag is on.

Note also that the logging is different between the two methods. The different logging is important to detect if we initialized synchronously or asynchronously.

The way I see to refactor this would lead to:
```
public final void initialize(boolean initializeAsync) {
        assertNeverCalledInitialize();

        if (!initializeAsync) {
            tryInitializeInternal();
            return;
        }

        tryAndCatchInitialization("Initialized {} synchronously.",
                "Failed to initialize {} in the first attempt, will initialize asynchronously.");
    }

    // Not final for tests.
    void scheduleInitialization() {
        singleThreadedExecutor.schedule(() -> {
            if (canceledInitialization) {
                return;
            }

            tryAndCatchInitialization("Initialized {} asynchronously.", "Failed to initialize {} asynchronously.");
        }, sleepIntervalInMillis(), TimeUnit.MILLISECONDS);
    }

    private void tryAndCatchInitialization(String messageWhenInitializedSuccessfully, String messageWhenFailedToInitialize) {
        try {
            tryInitializeInternal();
            log.info(messageWhenInitializedSuccessfully, SafeArg.of("className", getInitializingClassName()));
        } catch (Throwable throwable) {
            log.info(messageWhenFailedToInitialize,
                    SafeArg.of("className", getInitializingClassName()), throwable);
            cleanUpOnInitFailure();
            scheduleInitialization();
        }
    }
```
Which I think is less readable. What do you think?

Reply:
How important is it to distinguish these cases in the logs? As an alternative you could maintain a count of attempts and just log how many attempts it took (might also be interesting to know how long it took)

Reply:
It's going to make it the logs less clear (the behavior is not very explicit), but will make the code simpler. Done.

Do we also want to log the initialization time as a metric? Do we want to warn if initialization takes longer than a specific period?